boards/p1am_100/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+        }
+    }
+    PA09 {
+        name: d12,
+        aliases: {
+            AlternateC: Scl
+        }
+    }
+    PA10 {
+        name: d2,
+        aliases: {
+            AlternateG: I2sSerialClock
+        }
+    }
+    PA11 { name: d3 }
+    PA12 {
+        name: pa12,
+        aliases: {
+            AlternateC: SdSdo
+        }
+    }
+    PA13 {
+        name: pa13,
+        aliases: {
+            AlternateC: SdSck
+        }
+    }
+    PA14 {
+        name: pa14,
+        aliases: {
+            PushPullOutput: SdSlaveSelect
+        }
+    }
+    PA15 {
+        name: pa15,
+        aliases: {
+            AlternateC: SdSdi
+        }
+    }
+    PA16 {
+        name: d8,
+        aliases: {
+            AlternateC: Spi0Mosi
+        }
+    }
+    PA17 {
+        name: d9,
+        aliases: {
+            AlternateC: Spi0Sck
+        }
+    }
+    PA18 {
+        name: pa18,
+        aliases: {
+            /// Host Enable, drive high to switch into USB host mode
+            #[cfg(feature = "usb")]
+            PushPullOutput: UsbId
+        }
+    }
+    PA19 {
+        name: d10,
+        aliases: {
+            AlternateC: Spi0Miso
+        }
+    }
+    PA20 { name: d6 }
+    PA21 { name: d7 }
+    PA22 { name: d0 }
+    PA23 { name: d1 }
+    PA24 {
+        name: usb_dm,
+        aliases: {
+            #[cfg(feature = "usb")]
+            AlternateG: UsbDm
+        }
+    }
+    PA25 {
+        name: usb_dp,
+        aliases: {
+            #[cfg(feature = "usb")]
+            AlternateG: UsbDp
+        }
+    }
+    PA27 {
+        name: pa27,
+        aliases: {
+            PullUpInput: SdCardDetect
+        }
+    }
+    PA28 {
+        name: switch,
+        aliases: {
+            PullUpInput: Switch
+        }
+    }
+    PB02 {
+        name: d16,
+        aliases: {
+            AlternateB: A1
+        }
+    }
+    PB03 {
+        name: d17,
+        aliases: {
+            AlternateB: A2
+        }
+    }
+    PB08 {
+        name: led,
+        aliases: {
+            PushPullOutput: Led
+        }
+    }
+    PB09 {
+        name: pb09,
+        aliases: {
+            PushPullOutput: BaseEnable,
+            AlternateB: AdcBattery
+        }
+    }
+    PB10 { name: d4 }
+    PB11 { name: d5 }
+    PB22 {
+        name: d14,
+        aliases: {
+            AlternateD: UartTx
+        }
+    }
+    PB23 {
+        name: d13,
+        aliases: {
+            AlternateD: UartRx
+        }
+    }
+);
+
+const BASE_CONTROLLER_FREQ: Hertz = Hertz(1000000);
+// FIXME: const BASE_CONTROLLER_SPI_MODE: dyn spi::Mode = &spi::MODE_2;
+
+/// Convenience for setting up the labeled SPI0 peripheral.
+/// SPI0 has the P1AM base controller connected.
+/// This powers up SERCOM1 and configures it for talking to the
+/// base controller.
+pub fn base_controller_spi(
+    clocks: &mut GenericClockController,
+    sercom1: pac::SERCOM1,
+    pm: &mut pac::PM,
+    sck: Spi0Sck,
+    mosi: Spi0Mosi,
+    miso: Spi0Miso,
```

Comment:
This is a perfectly valid way to implement it. I just wanted to point out my logic for using `AnyPin`.

I once read a blog post or something that had advice about the types in a function signature. It argued that the input parameters should be typed as generically as possible, while the return values should be typed as specifically as possible. The idea is that you impose as few requirements as possible on the caller, and you provide as much information as possible about your result.

By using aliases here, you are requiring the pin to already be in the correct mode. That's fine, and it actually may produce easier to understand code. But it also means the caller is forced to convert the pin mode manually. With `impl AnyPin<Id = PXXX>`, the caller can provide a pin in any mode, so long as it is the right `PinId`.

Reply:
I greatly prefer being explicit with the caller (and it's just a `.into()` call for them). Especially since the pin names don't match their function, so the `.into()` is an explicit clue that something is being converted.